rules/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector.php

```diff
@@ -169,6 +170,18 @@ private function shouldSkipInferredReturnNode(ClassMethod | Function_ $functionL
         return (bool) $functionLike->returnType->getAttribute(AttributeKey::DO_NOT_CHANGE);
     }
 
+    private function hasUnionWithThisType(UnionType $unionType): bool
+    {
+        $types = $unionType->getTypes();
+        foreach ($types as $type) {
+            if ($type instanceof ThisType) {
+                return true;
+            }
+        }
+
+        return false;
+    }
```

Comment:
IMO this would skip any union type with `ThisType`. Instead, the `ThisType` should match the existing object type and compare it.

If they are equal, it shold be skipped. If they are no the same, they type should be overriden.

These 2 fixtures should be covered and work too:

```diff
class NullableSelf
{
    private ?self $parent = null;

-    public function getParent(): ?int
+   public function getParent(): ?self
    {
        return $this->parent;
    }
}
?>
```


```diff
class NullableSelf
{
    private ?self $parent = null;

-    public function getParent(): ?AnotherObject
+   public function getParent(): ?self
    {
        return $this->parent;
    }
}
?>
```

Reply:
@TomasVotruba I think those mostly a userbase invalid code which possibly on purpose. I tried that, but that cause invalid changes in other places due to check return expr and the return stmts as well.

Reply:
It might be the case, but this asumption might be wrong. It is rather about fixing the self comparison object that now does not work.

> I tried that, but that cause invalid changes in other places due to check return expr and the return stmts as well.

It should be fixed there too then. Could you send this proposal?

Reply:
I'm trying to remove this method to see the change, gimme a sec...